src/compiler/parser/index.js

```diff
@@ -201,6 +201,33 @@ export function parse (
     }
   }
 
+  function checkSlotConstraints (el: any, scopedSlots) {
+    scopedSlots = scopedSlots || {}
+    const children = el.children || []
+    if (!children.length) return
+    for (let i = 0; i < children.length; i++) {
+      const child = children[i]
+      if (child.tag !== 'slot') {
+        checkSlotConstraints(child, scopedSlots)
+        continue
+      }
+      const slotName = child.slotName || 'default'
+      const isScopedSlot = (child.attrs || []).concat(child.dynamicAttrs || []).map(attr => ({
+        name: camelize(attr.name),
+        value: attr.value,
+        dynamic: attr.dynamic
+      })).length > 0
+      if (isScopedSlot) {
+        scopedSlots[slotName] = true
+        continue
+      } else if (process.env.NODE_ENV !== 'production' && scopedSlots[slotName]) {
```

Comment:
Even though the check is at compiler level, I think it will make sense to wrap the check so it only gets executed on non-prod mod